adapter/adapter.py

```diff
@@ -0,0 +1,98 @@
+from boto.s3.connection import S3Connection, OrdinaryCallingFormat
+from boto.s3.bucket import Bucket
+from boto.s3.bucketlistresultset import BucketListResultSet
+from boto.exception import S3ResponseError
+from util import CephUtil
+from typing import List, Dict, Optional
+import leveldb
+
+class CephAdapter:
+
+    def __init__(self) -> None:
+        self._host = None
+        self._port = None
+        self._default_bucket_name = None
+        self.default_bucket = None
+
+    ### Internal methods
+
+    def _list_bucket(self) -> List[Bucket]:
+        try:
+            return self.conn.get_all_buckets()
+        except S3ResponseError as e:
+            raise CephAdapterError(e)
+
+    def _list_content(self) -> BucketListResultSet:
+        if self.default_bucket != None:
+            return self.default_bucket.list()
+        else:
+            raise CephAdapterError("Bucket is None")
+    
+    
+    ### Opened methods for ceph interaction
+
+    def connect(self, host:str, port:int, keyfile_path:str, default_bucket_name:str) -> None:
+        
+        self._host = host
+        self._port = port
+        self._default_bucket_name = default_bucket_name
+
+        try:
+            key_tuple = CephUtil.read_keys_from_file(keyfile_path)
+            self.conn = S3Connection(
+                aws_access_key_id = key_tuple[0],
+                aws_secret_access_key = key_tuple[1],
+                host = self._host,
+                port = self._port,
+                is_secure = False,
+                calling_format = OrdinaryCallingFormat(),
+            )
+
+            self.default_bucket = self.conn.get_bucket(self._default_bucket_name)
+
+        except EnvironmentError:
+            raise CephAdapterError("Read Keyfile error")
+        except S3ResponseError as s3err:
+            raise CephAdapterError("Ceph S3 error: "+ str(s3err))
+        except Exception as e:
+            raise CephAdapterError("Ceph Connect Error: "+str(e))
+
+    def upload(self, parcel_id: str, data: bytes) -> None:
+        try:
+            if self.default_bucket != None:
+                key = self.default_bucket.new_key(parcel_id)
+                key.set_contents_from_string(data)
+            else:
+                raise CephAdapterError("Bucket is None")
+        except ValueError as e:
+            raise CephAdapterError("Ceph Upload error: "+str(e))
+    
+    def download(self, parcel_id: str) -> bytes:
+        try:
+            if self.default_bucket != None:
+                key = self.default_bucket.get_key(parcel_id)
+                if key == None:
+                    raise CephAdapterError("Invalid Parcel Id")
```

Comment:
Shouldn't it be NOTFOUND or similar thing?

Reply:
I pushed the modified code with your comment reflected.